homeassistant/components/media_player/dlna_dmr.py

```diff
@@ -0,0 +1,693 @@
+        title_xml = root.find('.//dc:title', NS)
+        if title_xml is None:
+            return None
+
+        return title_xml.text
+
+    @property
+    @requires_state_variable('AVT', 'CurrentTrackMetaData')
+    def media_image_url(self, state_variable):
+        """Image url of current playing media."""
+        # pylint: disable=arguments-differ
+        xml = state_variable.value
+        if not xml:
+            return None
+
+        root = ET.fromstring(xml)
+        for res in root.findall('.//didl_lite:res', NS):
+            protocol_info = res.attrib.get('protocolInfo') or ''
+            if protocol_info.startswith('http-get:*:image/'):
+                url = protocol_info.text
+                return url
+
+        return None
+
+    @property
+    def state(self):
+        """State of the player."""
+        if not self._is_connected:
+            return STATE_OFF
+
+        avt_service = self._service('AVT')
+        if not avt_service:
+            return STATE_ON
+
+        transport_state = avt_service.state_variable('TransportState')
+        if not transport_state:
+            return STATE_ON
+        elif transport_state.value == 'PLAYING':
+            return STATE_PLAYING
+        elif transport_state.value == 'PAUSED_PLAYBACK':
+            return STATE_PAUSED
+
+        return STATE_IDLE
+
+    @property
+    @requires_state_variable('AVT', 'CurrentTrackDuration')
+    def media_duration(self, state_variable):
+        """Duration of current playing media in seconds."""
+        # pylint: disable=arguments-differ
+        if state_variable is None or state_variable.value is None:
+            return None
+
+        split = [int(v) for v in re.findall(r"[\w']+", state_variable.value)]
+        delta = timedelta(hours=split[0], minutes=split[1], seconds=split[2])
+        return delta.seconds
+
+    @property
+    @requires_state_variable('AVT', 'RelativeTimePosition')
+    def media_position(self, state_variable):
+        """Position of current playing media in seconds."""
+        # pylint: disable=arguments-differ
+        if state_variable is None or state_variable.value is None:
+            return None
+
+        split = [int(v) for v in re.findall(r"[\w']+", state_variable.value)]
+        delta = timedelta(hours=split[0], minutes=split[1], seconds=split[2])
+        return delta.seconds
+
+    @property
+    @requires_state_variable('AVT', 'RelativeTimePosition')
+    def media_position_updated_at(self, state_variable):
+        """When was the position of the current playing media valid.
+
+        Returns value from homeassistant.util.dt.utcnow().
+        """
+        # pylint: disable=arguments-differ
+        return state_variable.updated_at
+
+    @property
+    def name(self):
+        """Return the name of the device."""
+        return self._name
+
+    @property
+    def unique_id(self) -> str:
+        """Return an unique ID."""
+        return "{}.{}".format(__name__, self._url)
```

Comment:
This bugged me before and I'm glad you brought it up. The UUID is not known during the initial lifetime of the component. I.e., the component is initialized and on the first update the description file is parsed and the UUID becomes known.

Thus, the component might change unique_ids during its lifetime from None to a value. Is this acceptable for Home Assistant?

When the device is found via discovery, the same holds. Once the description file is parsed, the component knows the UUID.

Reply:
I'm not sure if it's okay to change the UUID after the initialization, as I think it will just get asked once during/after the setup phase. For non-discovered devices you could setup the device first after its service description is parsed, for discovered devices it is much simpler (as netdisco already parses the file during the discovery, and it could thus pass it onwards if not done already).